nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,530 @@
+    cmd_vel_in_topic, 1,
+    std::bind(&CollisionMonitor::cmdVelInCallback, this, std::placeholders::_1));
+  cmd_vel_out_pub_ = this->create_publisher<geometry_msgs::msg::Twist>(
+    cmd_vel_out_topic, 1);
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionMonitor::on_activate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Activating");
+
+  // Activating lifecycle publishers
+  cmd_vel_out_pub_->on_activate();
+
+  // Activating polygons
+  for (std::shared_ptr<PolygonBase> polygon : polygons_) {
+    polygon->activate();
+  }
+
+  // Since polygons are being published when cmd_vel_in appears,
+  // we need to publish polygons first time to display them at startup
+  publishPolygons();
+
+  // Activating main worker
+  process_active_ = true;
+
+  // Creating bond connection
+  createBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionMonitor::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+
+  // Deactivating main worker
+  process_active_ = false;
+
+  // Reset opeartion states to their defaults after worker deactivating
+  release_operation_ = false;
+  robot_action_prev_ = {DO_NOTHING, {0.0, 0.0, 0.0}};
+
+  // Deactivating polygons
+  for (std::shared_ptr<PolygonBase> polygon : polygons_) {
+    polygon->deactivate();
+  }
+
+  // Deactivating lifecycle publishers
+  cmd_vel_out_pub_->on_deactivate();
+
+  // Destroying bond connection
+  destroyBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionMonitor::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+
+  cmd_vel_in_sub_.reset();
+  cmd_vel_out_pub_.reset();
+  odom_sub_.reset();
+  first_odom_ = false;
+  velocity_received_ = false;
+
+  polygons_.clear();
+  sources_.clear();
+
+  tf_listener_.reset();
+  tf_buffer_.reset();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionMonitor::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+void CollisionMonitor::odomCallback(nav_msgs::msg::Odometry::ConstSharedPtr msg)
+{
+  // Obtaining odom_frame_id_ from first message and setting it as fixed frame for data sources
+  if (!first_odom_) {
+    for (std::shared_ptr<SourceBase> source : sources_) {
+      source->setFixedFrameId(msg->header.frame_id);
+    }
+    first_odom_ = true;
+  }
+
+  velocity_.x = msg->twist.twist.linear.x;
+  velocity_.y = msg->twist.twist.linear.y;
+  velocity_.tw = msg->twist.twist.angular.z;
+  velocity_stamp_ = msg->header.stamp;
+  velocity_received_ = true;
+}
+
+bool CollisionMonitor::velocityValid(const rclcpp::Time & curr_time)
+{
+  // Check whether velocity is not received yet
+  if (!velocity_received_) {
```

Comment:
Log a warning if we are checking for a velocity being valid and we have not yet received one. That could be diagnostic

Reply:
Since this is only used in the main thread for processing cmd_vels

Reply:
Where? That link doesn't render for me and it looks like that entire fucntion is gone. Specify the code block or lines

Reply:
It looks like rendering problem when it is being referenced to the file initially hidden (due to its large size).
The code block now moved to the main `process()` routine in a `collision_monitor_node.cpp` and as follows:
```
  // Warn if odometry was not received yet and do nothing
  if (odom_frame_id_.length() == 0) {
    RCLCPP_WARN(get_logger(), "Odometry is not received yet");
    return;
  }
```

`odom_frame_id_` is being set at the odometry subscriber once the first message was arrived.